File: research/experiments/sweep/runs/opt-large-csharp-sonnet45/Optimization/FiniteHessian.cs
namespace Optimization;

/// <summary>
/// Hessian matrix estimation via finite differences.
/// </summary>
public static class FiniteHessian
{
    private const double EPS = 2.22e-16; // Machine epsilon

    /// <summary>
    /// Compute the full n x n Hessian matrix using central differences.
    /// Step size: h = eps^(1/4) * max(|x_i|, 1) ≈ 1.22e-4
    /// Cost: 1 + 2n + n*(n-1) function evaluations.
    /// </summary>
    public static double[][] FiniteDiffHessian(Func<double[], double> f, double[] x)
    {
        int n = x.Length;
        double[][] H = new double[n][];
        for (int i = 0; i < n; i++)
        {
            H[i] = new double[n];
        }

        double fx = f(x);
        double h0 = Math.Pow(EPS, 0.25);

        // Diagonal elements
        for (int i = 0; i < n; i++)
        {
            double h = h0 * Math.Max(Math.Abs(x[i]), 1.0);
            double[] xPlus = VecOps.Clone(x);
            double[] xMinus = VecOps.Clone(x);
            xPlus[i] += h;
            xMinus[i] -= h;
            double fPlus = f(xPlus);
            double fMinus = f(xMinus);
            H[i][i] = (fPlus - 2 * fx + fMinus) / (h * h);
        }

        // Off-diagonal elements (upper triangle only, then mirror)
        for (int i = 0; i < n; i++)
        {
            double hi = h0 * Math.Max(Math.Abs(x[i]), 1.0);
            for (int j = i + 1; j < n; j++)
            {
                double hj = h0 * Math.Max(Math.Abs(x[j]), 1.0);

                double[] xpp = VecOps.Clone(x);
                xpp[i] += hi;
                xpp[j] += hj;

                double[] xpm = VecOps.Clone(x);
                xpm[i] += hi;
                xpm[j] -= hj;

                double[] xmp = VecOps.Clone(x);
                xmp[i] -= hi;
                xmp[j] += hj;

                double[] xmm = VecOps.Clone(x);
                xmm[i] -= hi;
                xmm[j] -= hj;

                double fpp = f(xpp);
                double fpm = f(xpm);
                double fmp = f(xmp);
                double fmm = f(xmm);

                H[i][j] = (fpp - fpm - fmp + fmm) / (4 * hi * hj);
                H[j][i] = H[i][j]; // Symmetry
            }
        }

        return H;
    }

    /// <summary>
    /// Approximate H*v using finite differences of the gradient:
    /// Hv ≈ (grad(x + h*v) - grad(x)) / h
    /// Step size: h = eps^(1/4) * max(||v||, 1)
    /// Cost: 1 gradient evaluation (gx is provided by caller).
    /// </summary>
    public static double[] HessianVectorProduct(
        Func<double[], double[]> grad,
        double[] x,
        double[] v,
        double[] gx)
    {
        double h = Math.Pow(EPS, 0.25) * Math.Max(VecOps.Norm(v), 1.0);
        double[] xPlusHv = VecOps.AddScaled(x, v, h);
        double[] gPlusHv = grad(xPlusHv);
        return VecOps.Scale(VecOps.Sub(gPlusHv, gx), 1.0 / h);
    }
}

File: research/experiments/sweep/runs/opt-large-go-sonnet45/finite_hessian.go
package optimization

import "math"

var fourthRootEps = math.Pow(2.220446049250313e-16, 0.25)

// FiniteDiffHessian computes the full Hessian matrix using central differences
func FiniteDiffHessian(f func([]float64) float64, x []float64) [][]float64 {
	n := len(x)
	H := make([][]float64, n)
	for i := range H {
		H[i] = make([]float64, n)
	}

	fx := f(x)
	h := make([]float64, n)
	for i := 0; i < n; i++ {
		h[i] = fourthRootEps * math.Max(math.Abs(x[i]), 1.0)
	}

	// Compute diagonal elements
	for i := 0; i < n; i++ {
		xPlus := Clone(x)
		xMinus := Clone(x)
		xPlus[i] += h[i]
		xMinus[i] -= h[i]

		fPlus := f(xPlus)
		fMinus := f(xMinus)

		H[i][i] = (fPlus - 2*fx + fMinus) / (h[i] * h[i])
	}

	// Compute off-diagonal elements (upper triangle)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			xPP := Clone(x)
			xPM := Clone(x)
			xMP := Clone(x)
			xMM := Clone(x)

			xPP[i] += h[i]
			xPP[j] += h[j]

			xPM[i] += h[i]
			xPM[j] -= h[j]

			xMP[i] -= h[i]
			xMP[j] += h[j]

			xMM[i] -= h[i]
			xMM[j] -= h[j]

			fPP := f(xPP)
			fPM := f(xPM)
			fMP := f(xMP)
			fMM := f(xMM)

			H[i][j] = (fPP - fPM - fMP + fMM) / (4 * h[i] * h[j])
			H[j][i] = H[i][j] // symmetry
		}
	}

	return H
}

// HessianVectorProduct computes H*v using finite differences of the gradient
func HessianVectorProduct(
	grad func([]float64) []float64,
	x, v []float64,
	gx []float64,
) []float64 {
	vNorm := Norm(v)
	h := fourthRootEps * math.Max(vNorm, 1.0)

	xPerturbed := AddScaled(x, v, h)
	gPerturbed := grad(xPerturbed)

	// (grad(x + h*v) - grad(x)) / h
	result := make([]float64, len(x))
	for i := range result {
		result[i] = (gPerturbed[i] - gx[i]) / h
	}

	return result
}
